golang extension: implement cluster_specifier_plugin to choose cluster dynamically

doujiang24 opened the pull request:
Commit Message:
golang extension: implement cluster_specifier_plugin to choose cluster dynamically
fix #25372 

Additional Description:
1. we could choose Host inside a cluster by using dynamic metadata, see #25371 ; but sometimes, we need to select a different cluster, i.e. dynamic_forward_proxy cluster vs STATIC cluster.
2. yep, `cluster_header` could be a way to reach it. but, we have to remove the hacked header if it's not expected from upstream.

so, adding a new plugin could be a proper choice.

Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

doujiang24 commented:
/review @wangfakang

phlax commented:
finally! the dep-review CI has shown itself useful!

```console
  contrib/golang/filters/http/test/test_data/echo/go.mod » golang.org/x/net@0.0.0-20190311183353-d8887717615a – golang.org/x/net/http2 Denial of Service vulnerability (high severity)
    ↪ https://github.com/advisories/GHSA-69cg-p879-7622
  contrib/golang/filters/http/test/test_data/echo/go.mod » golang.org/x/net@0.0.0-20190311183353-d8887717615a – Uncontrolled Resource Consumption (low severity)
    ↪ https://github.com/advisories/GHSA-vvpx-j8f3-3w6h
  contrib/golang/http/cluster_specifier/test/test_data/simple/go.mod » golang.org/x/net@0.0.0-20190311183353-d8887717615a – golang.org/x/net/http2 Denial of Service vulnerability (high severity)
    ↪ https://github.com/advisories/GHSA-69cg-p879-7622
  contrib/golang/http/cluster_specifier/test/test_data/simple/go.mod » golang.org/x/net@0.0.0-20190311183353-d8887717615a – Uncontrolled Resource Consumption (low severity)
    ↪ https://github.com/advisories/GHSA-vvpx-j8f3-3w6h
```

https://github.com/envoyproxy/envoy/actions/runs/4240541310/jobs/7369672618#step:3:16

wangfakang commented on api/contrib/envoy/extensions/http/cluster_specifier/golang/v3alpha/golang.proto:
```diff
@@ -0,0 +1,51 @@
+// [#extension-category: envoy.router.cluster_specifier_plugin]
+message Config {
+  // Globally unique ID for a dynamic library file.
+  string library_id = 1 [(validate.rules).string = {min_len: 1}];
```
We can separate it into a common type, which can be reused with the Go extension API. such as
```
message Library {
string id = 1 [(validate.rules).string = {min_len: 1}];
string path = 2 [(validate.rules).string = {min_len: 1}];
}
```

doujiang24 replied:
I prefer to keep it simple now since we could only reuse two fields, maybe we could turn to reuse it when we move dso into bootstrap.

wangfakang commented on contrib/golang/common/go/registry/BUILD:
```diff
@@ -0,0 +1,16 @@
+    deps = [
+        "//contrib/golang/filters/http/source/go/pkg/api",
+        "//contrib/golang/filters/http/source/go/pkg/http",
+        "//contrib/golang/http/cluster_specifier/source/go/pkg/cluster_specifier",
```
format

doujiang24 commented on contrib/golang/filters/http/test/test_data/passthrough/go.mod:
```diff
@@ -2,6 +2,8 @@ module example.com/passthrough
 go 1.18
 
-require github.com/envoyproxy/envoy v1.24.0
+require github.com/envoyproxy/envoy/contrib/golang v1.24.0
```
I prefer to keep the new changes since the old version doesn't work for now, we'd better make it work for now. And change it when the extension moves to standard.

Since the test cases may also be examples for users, making it work for now could be a better choice.
Users also have to change it when the extension moves to standard.

wangfakang commented on contrib/golang/http/cluster_specifier/source/cgo.cc:
```diff
@@ -0,0 +1,43 @@
+  auto goStr = reinterpret_cast<GoString*>(str);
+  return absl::string_view(goStr->p, goStr->n); // NOLINT(modernize-return-braced-init-list)
+}
+
```
add `#ifdef __cplusplus`

same elsewhere

wangfakang commented on contrib/golang/http/cluster_specifier/source/go/pkg/api/cluster.go:
```diff
@@ -0,0 +1,38 @@
+type ClusterSpecifierConfigFactory func(any *anypb.Any) ClusterSpecifier
+
+type RequestHeaderMap interface {
+	Get(key string) string
```
Get(key string) (string, bool)

wangfakang commented on contrib/golang/http/cluster_specifier/source/golang_cluster_specifier.cc:
```diff
@@ -0,0 +1,93 @@
+      dlib->envoyGoOnClusterSpecify(plugin_ptr, header_ptr, plugin_id, buffer_ptr, buffer_len);
+
+  if (new_len == 0) {
+    ENVOY_LOG(info, "golang cluster specifier choose the default cluster");
```
info -> debug

same elsewhere

wangfakang commented on contrib/golang/http/cluster_specifier/source/golang_cluster_specifier.cc:
```diff
@@ -0,0 +1,93 @@
+  auto new_len =
+      dlib->envoyGoOnClusterSpecify(plugin_ptr, header_ptr, plugin_id, buffer_ptr, buffer_len);
+
+  if (new_len == 0) {
```
new_len <= 0

wangfakang commented on contrib/golang/http/cluster_specifier/source/golang_cluster_specifier.cc:
```diff
@@ -0,0 +1,93 @@
+    cluster = std::string{buffer.data(), size_t(new_len)};
+  } else {
+    ENVOY_LOG(debug, "need larger size of buffer to save the cluster name in golang, try again");
+    buffer_len = new_len;
```
This needs to be protected, otherwise it will be a memory security vulnerability. Or directly limit a maximum length.

wangfakang reviewed (commented):
Some comments need to be addressed, others LGTM

doujiang24 commented:
/retest

wangfakang commented on contrib/golang/common/dso/dso.h:
```diff
@@ -17,54 +17,94 @@ class DsoInstance {
+                                  GoUint64 buffer_ptr, GoUint64 buffer_len);
+  GoUint64 envoyGoClusterSpecifierNewPlugin(GoUint64 config_ptr, GoUint64 config_len);
+
+protected:
```
```suggestion
private:
```

wangfakang commented on contrib/golang/common/dso/dso.cc:
```diff
@@ -53,9 +26,18 @@ DsoInstance::DsoInstance(const std::string dso_name) : dso_name_(dso_name) {
+}
+
+DsoInstance::~DsoInstance() {
+  // TODO(doujiang24): make sure the dso file is not no other users,
```
shared_ptr is used, so i don't think it's a issue.

doujiang24 replied:
yep, you are right, also added comments for it, thanks.

wangfakang commented on contrib/golang/common/dso/dso.h:
```diff
@@ -17,54 +17,94 @@ class DsoInstance {
   ~DsoInstance();
+  bool loaded() { return loaded_; }
+
+protected:
```
```suggestion
private:
```

wangfakang commented on contrib/golang/common/dso/dso.h:
```diff
@@ -17,54 +17,94 @@ class DsoInstance {
+
+protected:
+  const std::string dso_name_;
+  void* handler_{nullptr};
```
`handler_` It can only be read-only externally, and now becomes writable in the derived class. So it is safer to provide a function for export it. such as:
```
void* handler() {
    return handler_;
}

void setLoad(bool loaded) { loaded_ = loaded; }
```

doujiang24 replied:
I think we keep the same safety for `HttpFilterDsoInstance`, that no one could update `handler_` through `HttpFilterDsoInstance`'s API.

For `DsoInstance`, it's their responsibility to take care of `handle_` those who are inheriting `DsoInstance`, as well as using the private member.

So, I prefer using `protected` here.

wangfakang commented on contrib/golang/common/dso/dso.h:
```diff
@@ -17,54 +17,94 @@ class DsoInstance {
+                                           GoUint64 buffer_len) = {nullptr};
+};
+
 using DsoInstancePtr = std::shared_ptr<DsoInstance>;
```
remove unused type `DsoInstancePtr`.

doujiang24 commented:
/retest

wangfakang reviewed (dismissed):
LGTM and ping @wbpcode

doujiang24 commented:
friendly ping @wbpcode 

@mattklein123 Could you please take a look too? since this includes the new cluster_specifiy_plugin API.

doujiang24 commented:
resolved the conflicts and friend ping, thanks.

doujiang24 commented:
friendly ping @wbpcode 

I know it's some big, and maybe we could review it step by step, i.e. start from the API proto?

wbpcode commented on contrib/golang/common/dso/dso.h:
```diff
@@ -0,0 +1,147 @@
+  virtual void envoyGoFilterOnHttpDestroy(httpRequest* p0, int p1) PURE;
+};
+
+class HttpFilterDsoInstance : public HttpFilterDso {
```
Prefer `HttpFilterDsoImpl`.

wbpcode commented on contrib/golang/common/dso/dso.h:
```diff
@@ -0,0 +1,147 @@
+  virtual GoUint64 envoyGoClusterSpecifierNewPlugin(GoUint64 config_ptr, GoUint64 config_len) PURE;
+};
+
+class ClusterSpecifierDsoInstance : public ClusterSpecifierDso {
```
ditto.

wbpcode commented on contrib/golang/common/dso/dso.h:
```diff
@@ -0,0 +1,147 @@
+  };
+
+private:
+  using DsoMapType = std::map<std::string, std::shared_ptr<T>>;
```
could `absl::flat_hash_map` be used here?

wbpcode commented on contrib/golang/router/cluster_specifier/source/cgo.cc:
```diff
@@ -0,0 +1,55 @@
+extern "C" {
+#endif
+
+int envoyGoClusterSpecifierGetHeader(unsigned long long header_ptr, void* key, void* value) {
```
nit: it would be better if there are some more detailed comments to tell the reader the meaning of this func and parameters.

wbpcode commented on contrib/golang/router/cluster_specifier/source/go/pkg/api/capi.go:
```diff
@@ -0,0 +1,23 @@
+
+package api
+
+type HttpCAPI interface {
+	HttpGetHeader(headerPtr uint64, key *string, value *string) bool
+	HttpLogError(pluginPtr uint64, msg *string)
+}
```
could these shared by the http filter and cluster_specifier?

wbpcode commented on contrib/golang/router/cluster_specifier/source/golang_cluster_specifier.cc:
```diff
@@ -0,0 +1,95 @@
+  auto dlib = config_->getDsoLib();
+  ASSERT(dlib != nullptr);
+
+again:
```
could simple `while` loop be used here to replace this `goto`?

wbpcode reviewed (commented):
Thanks for your great contributions. This is great and some minor comments are added to start.

Sorry I have no much background knowledge about the cgo and ffi. So, If I am wrong or misunderstand something, feel free to point it out.

wbpcode commented:
And seems there is a refactor to the dso manager and dso interface. Could we split these changes to separated PR？

doujiang24 commented on api/contrib/envoy/extensions/router/cluster_specifier/golang/v3alpha/golang.proto:
```diff
@@ -0,0 +1,51 @@
+
+// [#extension-category: envoy.router.cluster_specifier_plugin]
+message Config {
+  // Globally unique ID for a dynamic library file.
+  string library_id = 1 [(validate.rules).string = {min_len: 1}];
+
+  // Path to a dynamic library implementing the
+  // :repo:`ClusterSpecifier API <contrib/golang/router/cluster_specifier/source/go/pkg/api.ClusterSpecifier>`
+  // interface.
+  // [#comment:TODO(wangfakang): Support for downloading libraries from remote repositories.]
+  string library_path = 2 [(validate.rules).string = {min_len: 1}];
+
+  // Default cluster.
+  //
+  // It will be used when the specifier interface return empty string or panic.
+  //
+  string default_cluster = 3 [(validate.rules).string = {min_len: 1}];
+
+  // Configuration for the Go cluster specifier plugin.
+  //
+  // .. note::
+  //     This configuration is only parsed in the go cluster specifier, and is therefore not validated
+  //     by Envoy.
+  //
+  //     See the :repo:`StreamFilter API <contrib/golang/router/cluster_specifier/source/go/pkg/cluster_specifier/config.go>`
+  //     for more information about how the plugin's configuration data can be accessed.
+  //
+  google.protobuf.Any config = 4;
```
Not exactly, it's `plugin_config` in http filter,  I personally prefer to keep them separated, since they are both in the early stage, so that we could iterate them separately.

I'm fine if you think it's better to share the `library_id` and `library_path`, at this stage.

doujiang24 commented:
> And seems there is a refactor to the dso manager and dso interface. Could we split these changes to separated PR？

Split it to https://github.com/envoyproxy/envoy/pull/26116, please take a look there, thanks!

Thanks @wbpcode , all comments are addressed, please take another look, thanks!

doujiang24 commented:
friendly ping @wbpcode 

already resolved the merge conflicts, please take another look, thanks!

alyssawilk commented:
@wbpcode ping?

alyssawilk commented:
as a reminder @wbpcode AFIK maintainers aren't responsible for code quality in contrib, AFIK we just make sure CI is happy and contrib authors/reviewers are happy and you're good to merge.

wbpcode commented:
got it. then I will just do a quick check tomorrow.

wbpcode reviewed (dismissed):
LGTM. Thanks.

wbpcode commented:
cc @doujiang24 could you resolve the conflicts first then we can merge it before the weekend? 😄

doujiang24 commented:
@wbpcode Thanks, resolved conflicts, glad to see this could be merged soon :)

wbpcode commented:
> @wbpcode Thanks, resolved conflicts, glad to see this could be merged soon :)

Thanks for your create work.

wbpcode reviewed (approved):
LGTM. Thanks.
